Implemented applicant fetching

jmilic1 opened the pull request.

Commit:
Implemented applicant fetching

vpintaric-axilis commented on lecture_1/assignement/go.mod:
```diff
@@ -0,0 +1,8 @@
+module code-cadets-2021/lecture_1/02_yes_no_example
```
Module name doesn't really matter here since this tool isn't supposed to be imported by others but to keep it consistent this should be `module code-cadets-2021/lecture_1/assignement`.

Also note the typo -> not "assignement" - it should be "assignment".

vpintaric-axilis commented on lecture_1/assignement/main.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func fetchContent() []jobApiResponse {
+	httpClient := pester.New()
+	httpClient.Backoff = linearBackoff
```
Using `pester` is not wrong per se but is kind-of an overkill for this. Using regular http client from the standard library is enough.

```suggestion
	httpClient := &http.Client{}
```

jmilic1 replied:
Would it be appropriate to use http.Get() method to fetch the data since we do not need a stateful client for this task?

vpintaric-axilis replied:
Yep `http.Get()` works as well.

vpintaric-axilis commented on lecture_1/assignement/main.go:
```diff
@@ -0,0 +1,96 @@
+	httpClient.Backoff = linearBackoff
+
+	httpResponse, err := httpClient.Get(jobApi)
+	if err != nil {
+		log.Fatal(
+			errors.WithMessage(err, "HTTP get towards job_api API"),
+		)
+	}
```
It can be dangerous to handle errors like this in functions. This would abruptly end the program without the callers being aware of this. Generally functions should always return the error as a value. Once the error is received by the `main` method it can choose what to do with it.

vpintaric-axilis commented on lecture_1/assignement/main.go:
```diff
@@ -0,0 +1,96 @@
+		}
+
+		skills := formatSkills(val.Skills)
+		if strings.Contains(skills, "Go") || strings.Contains(skills, "Java") {
```
It would be better to iterate over the `Skills` array and check each element directly. Merging the array into a single string and then using `contains` could result in some weird bugs.

jmilic1 replied:
Is the implementation from new commits (which creates a set from given array) an OK solution or is there something a bit more elegant which can be used here?

vpintaric-axilis replied:
Yep, it's a bit of an overkill though since you're converting a slice into a set and then using it only once but I wouldn't worry about that.

vpintaric-axilis commented on lecture_1/assignement/main.go:
```diff
@@ -0,0 +1,96 @@
+
+func writeToFile(f *os.File, content []jobApiResponse) {
+	for _, val := range content {
+		if !val.Passed {
```
Note that this is not according to the specification! Applicants should be added to the file only if they both passed and if they have Go and/or Java as a skill.

jmilic1 replied:
Isn't this according to the specification though?

We check if the applicant passed or not. If they did not, we just skip over their entry altogether with `continue`.
If they did pass, we then check if they have Java or Go as their skill.

Am I missing something?

Also, I just noticed that this implementation will write an applicant that passed and knows only Javascript whoops

vpintaric-axilis replied:
Ah yeah, sorry, my bad.

And you've fixed the javascript issue 👌

Commit:
Removed assignement typo.

Renamed module.

Changed content fetching to use http.Get instead of a client.

Commit:
Changed error handling to propagate all errors to main.

Changed skill checking to first check if necessary skills exist and if they exist, format the skills into an output.

Commit:
Changed content fetching to use basic Golang client.

vpintaric-axilis reviewed (approved):
👌
